[aqi, hm3301, pmsx003] Air Quality Index improvements

jasstrong opened the pull request:
# What does this implement/fix?

 This PR adds Air Quality Index (AQI) calculation support to the pmsx003 component and refactors the existing hm3301 AQI implementation into a shared aqi component that both sensors (and future particulate matter sensors) can use.

## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Developer breaking change (an API change that could break external components)
- [x] Code quality improvements to existing code or addition of tests
- [x] Other

**Related issue or feature (if applicable):**
  - New shared aqi component with both EPA AQI (US) and CAQI (European) calculators
  - Added AQI calculation support to pmsx003 sensors
  - Refactored hm3301 to use the shared component, eliminating code duplication
  - Fixed pmsx003 sensor initialization to properly send active/passive mode commands
  - Memory-safe implementation of AQI computation with zero dynamic allocation (fixes a bare `new` in hm3301)

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):**

- I have added examples and inline documentation in this PR-  there's also a docs PR https://github.com/esphome/esphome-docs/pull/5703

## Test Environment

- [ ] ESP32
- [ ] ESP32 IDF
- [x] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx
- [ ] nRF52840

## Example entry for `config.yaml`:

```yaml

  # PMSX003 example
  uart:
    rx_pin: GPIO16
    tx_pin: GPIO17
    baud_rate: 9600

  sensor:
    - platform: pmsx003
      type: PMSX003
      pm_2_5:
        name: "PM2.5"
      pm_10_0:
        name: "PM10"
      aqi:
        name: "Air Quality Index"
        calculation_type: "AQI"  # or "CAQI" for European standard

  # HM3301 example (existing functionality preserved)
  i2c:
    sda: GPIO21
    scl: GPIO22

  sensor:
    - platform: hm3301
      pm_2_5:
        name: "PM2.5"
      pm_10_0:
        name: "PM10"
      aqi:
        name: "Air Quality Index"
        calculation_type: "CAQI"  # European CAQI standard
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

Commit:
pmsx003: Add AQI calculation and extract to shared component

This commit adds AQI (Air Quality Index) calculation support to the
pmsx003 component and refactors the existing hm3301 AQI implementation
into a shared component that both sensors can use.

Changes:
- Created new 'aqi' component with AQI/CAQI calculators and factory
- Extracted AQI calculation code from hm3301 into shared component
- Added AQI calculation support to pmsx003 sensor
- Both components now use the shared aqi component (AUTO_LOAD)
- Updated to use 2022 EPA AQI calculation standards
- Fixed pmsx003 sensor initialization to properly send active/passive
  mode commands on first loop iteration
- Added component tests for AQI feature
- Added example configurations and documentation

The AQI calculator supports both:
- AQI: US EPA Air Quality Index (0-500 scale)
- CAQI: European Common Air Quality Index (0-400 scale)

Both calculation types use PM2.5 and PM10 concentrations to compute
the overall air quality index, returning the higher of the two values.

Configuration example:
  sensor:
    - platform: pmsx003
      pm_2_5:
        name: "PM2.5"
      pm_10_0:
        name: "PM10"
      aqi:
        name: "AQI"
        calculation_type: "AQI"  # or "CAQI"

jasstrong committed:
Merge branch 'dev' into jas/aqi-for-pmsx003

andrewjswan commented:
Perhaps we should standardize this sensor, make it separate, remove it from all components, and simply specify the input data for it.

```Yaml
  sensor:
    - platform: aqi
      name: "Air Quality Index"
      calculation_type: "CAQI"  # European CAQI standard
      pm25: pm25_id
```

jasstrong commented:
can I get some reviews please?  it's been a few days.

swoboda1337 commented on esphome/components/aqi/abstract_aqi_calculator.h:
```diff
@@ -3,12 +3,12 @@
 
 namespace esphome {
-namespace hm3301 {
+namespace aqi {
```
```suggestion
namespace esphome::aqi {
```
Should update all namespaces and the comments

swoboda1337 commented on esphome/components/pmsx003/pmsx003.cpp:
```diff
@@ -248,6 +266,13 @@ void PMSX003Component::parse_data_() {
     this->pm_particles_25um_sensor_->publish_state(pm_particles_25um);
 
+  // Calculate AQI if sensor is configured
+  int aqi_value = -1;
```
```suggestion
  int32_t aqi_value = -1;
```

swoboda1337 commented on esphome/components/pmsx003/pmsx003.cpp:
```diff
@@ -271,6 +296,10 @@ void PMSX003Component::parse_data_() {
+  // Publish AQI value if calculated
+  if (aqi_value != -1 && this->aqi_sensor_ != nullptr)
+    this->aqi_sensor_->publish_state(aqi_value);
+
```
Why not move this into the if above?

Commit:
fixup! pmsx003: Add AQI calculation and extract to shared component

Address review feedback:
- Use combined namespace style (esphome::aqi)
- Use int32_t instead of int for aqi_value
- Move AQI publish into calculation block

jasstrong committed:
Merge branch 'dev' into jas/aqi-for-pmsx003

swoboda1337 commented on esphome/components/pmsx003/pmsx003.cpp:
```diff
@@ -248,6 +266,13 @@ void PMSX003Component::parse_data_() {
     this->pm_particles_25um_sensor_->publish_state(pm_particles_25um);
 
+  // Calculate and publish AQI if sensor is configured
+  if (this->aqi_sensor_ != nullptr && this->pm_2_5_sensor_ != nullptr && this->pm_10_0_sensor_ != nullptr) {
```
```suggestion
  if (this->aqi_sensor_ != nullptr) {
```
The values are still calculated even without these sensors being configured no point gating it

swoboda1337 commented on esphome/components/pmsx003/pmsx003.cpp:
```diff
@@ -18,6 +18,16 @@ static const uint16_t PMS_CMD_MEASUREMENT_MODE_ACTIVE = 0x0001;  // automaticall
+    ESP_LOGCONFIG(TAG, "Will use passive mode with sleep/wake cycles");
+  }
+}
+
```
Why not remove this and put this info in dump_config? No one really sees setup as you need serial logging.

jasstrong replied:
I used it while I was debugging the sensor being stuck in a mode where it didn't count particles.  I fixed that by resetting it more thoroughly.  I can just remove it since it's not really useful anyway.

swoboda1337 commented on esphome/components/pmsx003/sensor.py:
```diff
@@ -35,11 +36,18 @@
+UNIT_INDEX = "index"
+
+AQI_CALCULATION_TYPE = aqi.AQI_CALCULATION_TYPE
+
```
```suggestion
UNIT_INDEX = "index"
```
Can you just import CONF_AQI etc from aqi like we do for esphome.const, can remove these lines.

swoboda1337 commented on esphome/components/pmsx003/sensor.py:
```diff
@@ -192,6 +204,19 @@ def validate_update_interval(value):
                 state_class=STATE_CLASS_MEASUREMENT,
             ),
+            cv.Optional(CONF_AQI): sensor.sensor_schema(
+                unit_of_measurement=UNIT_INDEX,
```
```suggestion
```
There is no unit, should only use valid unit, users can add this in yaml if they want

jasstrong committed:
Update esphome/components/pmsx003/pmsx003.cpp

take @swoboda1337's suggestion

Co-authored-by: Jonathan Swoboda <154711427+swoboda1337@users.noreply.github.com>

jasstrong committed:
Update esphome/components/pmsx003/sensor.py

take @swoboda1337's suggestion

Co-authored-by: Jonathan Swoboda <154711427+swoboda1337@users.noreply.github.com>

jasstrong committed:
Update esphome/components/pmsx003/sensor.py

take @swoboda1337's suggestion

Co-authored-by: Jonathan Swoboda <154711427+swoboda1337@users.noreply.github.com>

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

Commit:
Move mode logging from setup() to dump_config()

Per review feedback: logging in setup() requires serial logging to see.
Move the active/passive mode information to dump_config() where it
will be visible in the standard configuration output.

Commit:
Import CONF_AQI and AQI_CALCULATION_TYPE from aqi component

swoboda1337 commented on esphome/components/pmsx003/sensor.py:
```diff
@@ -35,11 +37,13 @@
 PMSX003Component = pmsx003_ns.class_("PMSX003Component", uart.UARTDevice, cg.Component)
 PMSX003Sensor = pmsx003_ns.class_("PMSX003Sensor", sensor.Sensor)
 
+UNIT_INDEX = "index"
```
Doesn't seem to be used but I guess https://github.com/esphome/esphome/blob/dev/esphome/components/hm3301/sensor.py#L71 is already using UNIT_INDEX so its fine to set the unit here. Good to be consistent. Sorry about that.

swoboda1337 commented on esphome/components/hm3301/sensor.py:
```diff
@@ -16,22 +16,19 @@
-CONF_CALCULATION_TYPE = "calculation_type"
+CONF_AQI = aqi.CONF_AQI
+CONF_CALCULATION_TYPE = aqi.CONF_CALCULATION_TYPE
 UNIT_INDEX = "index"
```
Should just import like you are doing in the other file: 
```
from esphome.components.aqi import AQI_CALCULATION_TYPE, CONF_AQI, CONF_CALCULATION_TYPE
```

Commit:
Import AQI constants directly from aqi component in hm3301

Commit:
Add unit_of_measurement to AQI sensor to match hm3301

jasstrong committed:
Merge branch 'dev' into jas/aqi-for-pmsx003

jasstrong committed:
Merge branch 'dev' into jas/aqi-for-pmsx003

andrewjswan commented on esphome/components/aqi/aqi_calculator.h:
I think in this block it is also worth checking the result of `grid_index`, and if we don't have an index, just exit the function.

```suggestion
    int grid_index = get_grid_index_(value, array);
    if (grid_index == -1) {
      return -1;
    }
```

swoboda1337 replied:
That wasn't modified in this PR however yeah that code should be updated.

jasstrong replied:
yeah, I missed that.  sorry!

Commit:
Add grid_index bounds check in AQI calculator
